Pipelining dispatcher improvements

plaflamme opened the pull request:
Problem

`PipeliningDispatcher` has the following limitations:

* same type signature as the underlying `Transport`;
* hardcoded timeout for detecting a stalled pipeline;
* does not allow tracking more than the dispatched `Promise` in the pipeline.

Solution

Abstract over the dispatcher / transport types, add another type parameter for what is being tracked in the pipeline.

Add a new `Param` to allow configuring the stall timeout.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twitter/finagle?pullRequest=624) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/twitter/finagle?pullRequest=624) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twitter/finagle?pullRequest=624) it.</sub>

dschobel commented:
thanks for the pr @plaflamme, lgtm big picture. I'm curious what your use case is if you don't mind saying.

plaflamme commented on finagle-core/src/main/scala/com/twitter/finagle/dispatch/PipeliningDispatcher.scala:
```diff
@@ -54,10 +55,14 @@ private[finagle] class PipeliningDispatcher[Req, Rep](
+
   // Dispatch serialization is guaranteed by GenSerialClientDispatcher so we
   // leverage that property to sequence `q` offers.
   protected def dispatch(req: Req, p: Promise[Rep]): Future[Unit] =
```
I hesitated to do that and didn't similar to how `apply` is not final in `GenSerialClientDispatcher`. I figured that most use-cases would only need to override `pipeline`, but potentially more complex ones would need to override `dispatch` as well. I don't have a strong opinion, either way is fine by me.

plaflamme commented:
The use case is a `finagle-kafka` client which we may be able to contribute at some point. The reason why it needs these changes is mostly because Kafka has a `correlation id` associated with each request (basically a unique id) which is sent back to clients in the response. This is a dispatching concern, so I wanted to encapsulate that in there.

The `correlation id` is put into the queue along with the promise and is checked when the response comes from the server. It is required to match head of the queue and then simply thrown away.

In its current state, it also uses different types between the dispatcher and the transport, but that's probably not strictly necessary, just a nice to have.

The configuration of the stall timeout is strictly hygiene (I figured it would be better if it were configurable). So we can drop that commit if it's preferable to minimize the change.

plaflamme commented:
@dschobel Sorry, I had to rebase the original PR because it wasn't off `develop`.

I added an additional commit which I hope addresses your concerns.

It also enforces correctness of handling the `Promise[Rep]` which was originally left to extending classes. I think this is an improvement, but it came at the cost of instantiating an extra type in the simple case.

Also, an argument to leave `dispatch` non-final is that Kafka actually has an edge case where one particular request does not get a server response. So it would actually need to override `dispatch` to avoid enqueuing an item in the queue for those requests.

An alternative is to return an `Option[T]` from `pipeline` (or some other union type) that `dispatch` would know not to enqueue (say when it's `None`), but it seemed like too much of an edge case to add formal support for. Leaving `dispatch` non-final seemed simpler.

dschobel commented on finagle-core/src/main/scala/com/twitter/finagle/dispatch/PipeliningDispatcher.scala:
```diff
@@ -85,14 +109,14 @@ private[finagle] class PipeliningDispatcher[Req, Rep](
 
-private object PipeliningDispatcher {
+object GenPipeliningDispatcher {
   val log = Logger.get(getClass.getName)
```
`log` can stay private I believe

dschobel commented on finagle-core/src/main/scala/com/twitter/finagle/dispatch/PipeliningDispatcher.scala:
```diff
@@ -85,14 +109,14 @@ private[finagle] class PipeliningDispatcher[Req, Rep](
+  private case class Pending[T, Rep](value: T, promise: Promise[Rep])
 
-  val StalledPipelineException =
+  def stalledPipelineException(timeout: Duration) =
```
I think this one can be private as well

dschobel commented on finagle-core/src/main/scala/com/twitter/finagle/dispatch/PipeliningDispatcher.scala:
```diff
@@ -103,3 +127,23 @@ private object PipeliningDispatcher {
   }
 }
+
+case class StalledPipelineTimeout(timeout: Duration) {
```
docs please

dschobel commented on finagle-memcached/src/main/scala/com/twitter/finagle/Memcached.scala:
```diff
@@ -1,11 +1,12 @@
 package com.twitter.finagle
 
 import _root_.java.net.SocketAddress
+
```
rm space please

dschobel commented on finagle-memcached/src/main/scala/com/twitter/finagle/Memcached.scala:
```diff
@@ -29,6 +30,7 @@ import com.twitter.hashing
 import com.twitter.io.Buf
 import com.twitter.util.registry.GlobalRegistry
 import com.twitter.util.{Closable, Duration, Monitor}
+
```
ditto, no space

plaflamme commented:
@dschobel I've added a commit with the requested changes. I'll sign the CLA ASAP.

plaflamme commented:
@dschobel @kevinoliver any guidance on how to solve the test failure? Is it a flake?

dschobel commented:
@plaflamme looks like it was a flaky run. I'm merging the pr internally and I'll report back if it reappears.

dschobel commented:
merged in affced891e62428e156f31e2c14dd32121404582, thanks again for the pr @plaflamme
